cranelift-sizeless-vector.md

```diff
@@ -18,12 +18,80 @@ Rust is also currently starting to use LLVM's ScalableVectorType and, as a targe
+We can add sizeless vector types by modifying existing structs to hold an extra bit of information to represent the sizeless nature.
+
+## Type System
+- The new types do not report themselves as vectors, so ty.is\_vector() = false, but are explicitly reported via ty.is\_sizeless\_vector().
+- is\_vector is also renamed to is\_sized\_vector to avoid ambiguity.
 - The TypeSet and ValueTypeSet structs gain a bool to represent whether the type is sizeless.
 - TypeSetBuilder also gains a bool to control the building of those types.
-- At the encoding level, a bit is used to represent whether the type is sizeless, this bit has been taken from the max number of vector lanes supported, so they'd be reduced to 128 from 256.
+- At the encoding level, a bit is used to represent whether the type is sizeless and this bit has been taken from special types range.
+- These changes allow the usual polymorphic vector operations to be automatically built for the new set of sizeless types.
+
+## IR Entities and Function Changes
+A new entity is added to the IR, the SizelessStackSlot, and the Function will hold these separately from the existing StackSlot entities, also providing two APIs to create them:
+- create\_sized\_stack\_slot
+- create\_sizeless\_stack\_slot
+
+Keeping two vectors enables us to continue to use each entity's index as it's slot identifier, and allows us to place the entities in different positions in the frame. It also enables the backends to easily track which slots are sizeless, and which are not.
+
+## Instructions
+Three new instructions are added at the IR level to use the new SizelessStackSlot entity:
+- SizelessStackAddr
+- SizelessStackLoad
+- SizelessStackStore
+
+The primary difference between these operations and their existing counterparts is that they only take a SizelessStackSlot operand, without a byte offset.
+
+## ABI Layer Changes
+
+- The method stack_stackslot_addr is renamed to sized\_stackslot\_addr.
+- The method sizeless\_stackslots\_addr is introduced, which takes a vector\_scale parameter.
+- get\_number\_of\_spillslots\_for\_value is also modified to take a vector\_scale parameter.
+
+A key challenge to supporting these new types is that the register allocator expects to be given an constant value for the size of a spill slot, for a given register class. So, the current expectation is that the backends will continue to provide a fixed number, potentially larger that they currently do. This value is provided to the ABI layer via a new method on the TargetIsa trait, vector\_scale, which returns the largest number of bytes for a target vector register. This can then also be used to scale the index when calculating the address of a SizelessStackSlot, if a backend chooses a fixed sized during code generation.
+
+With the notion of a sizeless stack slot possible, but not a sizeless spill slot, the proposed frame layout would look like the following:
```

Comment:
Key question that remains for me here just from reading the RFC: at what time do we know the size (i.e., at compile time or only at runtime -- I think in discussion you had said it will be known at compile time?), and then how is this communicated to the ABI code?

Is the size of a sizeless vector actually always given by a method on the `TargetIsa`, and this is meant to be queried by the ABI code? Or does it have to be passed into the prologue somehow by e.g. giving it a global value that loads this config from a vmctx or register or...?

Reply:
The current proposal only gets us half way to the runtime length support, as the register allocator wants a fixed number for a the size of a spill slot for a given register class. I've tried to enable 'sizeless' as much as possible with the hope of playing with regalloc2 to enable full runtime support later. So, for now, TargetIsa would have to provide a fixed size and so TargetIsa is passed, instead of settings::Flags, when calling new on ABICalleeImpl.

Reply:
I'd add that modifying the backend ABI implementations to further support runtime lengths shouldn't be complicated but there is the issue of the frame size calculation for stack size checks and offset initialization, and I haven't looked into this much.